A workable version for Softlayer

mattcui opened the pull request:
Already verified with `zookeeper` release.

/cc @maximilien @jianqiu

cppforlife commented on softlayer/cpi.yml:
```diff
@@ -0,0 +1,98 @@
+    apiKey: ((sl_api_key))
+    featureOptions:
+      apiEndpoint: api.service.softlayer.com
+      apiRetryCount: 10
```
why arent these property values default on the cpi?

mattcui replied:
By default, it uses public SL API endpoint (api.softlayer.com), I just hope to show there is an option to change to use the private SL API endpoint, I could delete them if you don't think it's necessary.

cppforlife replied:
> I just hope to show there is an option to change to use the private SL API endpoint, I could delete them if you don't think it's necessary.

that should be covered by docs & https://bosh.io/jobs/softlayer_cpi?source=github.com/cloudfoundry/bosh-softlayer-cpi-release

mattcui replied:
OK, I am deleting them, and yes, it's already covered in https://bosh.io/jobs/softlayer_cpi?source=github.com/cloudfoundry/bosh-softlayer-cpi-release

cppforlife commented on softlayer/cpi.yml:
```diff
@@ -0,0 +1,98 @@
+  path: /instance_groups/name=bosh/properties/softlayer?
+  value: &softlayer
+    username: ((sl_username))
+    apiKey: ((sl_api_key))
```
keys should be snake case.

mattcui replied:
OK, I will modify it. Thanks.

cppforlife commented on softlayer/cloud-config.yml:
```diff
@@ -0,0 +1,51 @@
+    hourlyBillingFlag: true
+    vmNamePrefix: ((sl_vm_name_prefix))
+    domain: ((sl_vm_domain))
+    datacenter:
```
datacenter is already specified on the az, so no need to specify on vm type.

mattcui replied:
OK, I will have test to make sure our CPI works after removing `datacenter` from `vm_types`

cppforlife commented on softlayer/cloud-config.yml:
```diff
@@ -0,0 +1,51 @@
+azs:
+- name: z1
```
should have 3 azs even if as a starter they are all the same. makes default example manifests work much better.

mattcui replied:
OK, I will add more for Dal and SJ. Thanks.

cppforlife commented on softlayer/cloud-config.yml:
```diff
@@ -0,0 +1,51 @@
+vm_types:
+- name: default
+  cloud_properties:
+    bosh_ip: ((internal_ip))
```
why do we need internal_ip over here?

mattcui replied:
It's used to combine mbus url in user_data.json, we are already able to set the private IP for dynamic networking, but haven't implemented it for static networking.

cppforlife commented on softlayer/cloud-config.yml:
```diff
@@ -0,0 +1,51 @@
+    datacenter:
+      name: ((sl_datacenter))
+
+disk_types:
```
should have a disk type called `default`

mattcui replied:
Yes, I will modify. Thanks.

cppforlife commented on softlayer/cpi.yml:
```diff
@@ -0,0 +1,98 @@
+    ephemeralDiskSize: 100
+    startCpus: 4
+    maxMemory: 8192
+    deployedByBoshcli: true
```
why do we need `deployedByBoshcli`

mattcui replied:
Actually, it's used to get ride of `bosh_ip` from each resource pools (in manifest v1), for deploying bosh director, there needs to update /etc/hosts for supporting dynamic IP. Below is the code where to use it.

https://github.com/cloudfoundry/bosh-softlayer-cpi-release/blob/master/src/bosh-softlayer-cpi/softlayer/vm/softlayer_virtual_guest_creator.go#L103-L130

cppforlife commented:
@mattcui left some comments.

mattcui commented:
@cppforlife I addressed your comments, please review again, thank you. For comments you said there left, do you mean the comments like `# Add CPI job`, I saw it's in cpi.yml of all other IaaSes, do I need to removed all comments in cpi.yml? Thanks.

cppforlife commented on softlayer/cpi.yml:
```diff
@@ -0,0 +1,90 @@
+  path: /instance_groups/name=bosh/properties/softlayer?
+  value: &softlayer
+    username: ((sl_username))
+    apiKey: ((sl_apiKey))
```
this one seems to still have mixed casing case.

cppforlife commented:
i meant comments on github =)

mattcui commented:
@cppforlife I changed `sl_apiKey` to `sl_api_key`, please let me know if any other things I need to change. Thanks.

cppforlife commented:
I merged this PR; however, let's add softlayer config to `test.sh`. Also am i understanding correctly that /etc/hosts update isnt nececssary anymore?
